### PR TITLE
Restructuring SslHandler to support new engines

### PR DESCRIPTION
**Pull request by nmittler**
Motivation:

In preparation for support of Conscrypt, I'm consolidating all of the engine-specific details so that it's easier to add new engine types that affect the behavior of SslHandler.

Modifications:

Added an enum SslEngineType that provides SSL engine-specific details.

Result:

SslHandler is more extensible for other engine types.

### CODE REVIEW

**Comment by nmittler** [comment]
@normanmaurer can you take a look?

**Comment by normanmaurer** [comment]
Sure thing... still on my way back from Devoxx but will do

**Comment by nmittler** [comment]
Ok great ... this is a pretty easy one.  Just moving code around a bit ... no functionality changes.

**Comment by normanmaurer** [comment]
@nmittler actually I am not sure why we even need this ... We should just handle all the SSLEngine implementations the same way and only do optimisations  for "known" ones like the OpenSslEngine which is part of Netty itself. So I would prefer to not pull this in

**Comment by nmittler** [comment]
@normanmaurer I'm not sure I understand your point. The point of this PR is to consolidate the code for "known" engines into one place rather than having little hacks (for OpenSSL or coming soon ... conscrypt) scattered throughout the code, as is done in these places:

https://github.com/netty/netty/blob/4.1/handler/src/main/java/io/netty/handler/ssl/SslHandler.java#L277
https://github.com/netty/netty/blob/4.1/handler/src/main/java/io/netty/handler/ssl/SslHandler.java#L287
https://github.com/netty/netty/blob/4.1/handler/src/main/java/io/netty/handler/ssl/SslHandler.java#L1073

Consolidating the logic into the enum makes it very clear upfront as to how SslHandler handles the various known engines differently. And it also gets rid of extra conditionals at runtime.

**Comment by normanmaurer** [comment]
@nmittler actually never mind after talking with you on chat I now understand what you are doing... Ship it!

**Comment by normanmaurer** [comment]
@Scottmitch you want to check as well ?

**Comment by nmittler** [comment]
@Scottmitch gentle ping :)

**Comment by nmittler** [comment]
@normanmaurer I can't imagine the resource leak was caused by this PR ... can you confirm?

**Comment by Scottmitch** [comment]
@nmittler - Leak looks related to https://github.com/netty/netty/issues/5850#issuecomment-261073798

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java [hunk 1]
```diff
@@ -179,8 +179,89 @@
...
+                        toByteBuffer(out, writerIndex, out.writableBytes()));
+                }
+                out.writerIndex(writerIndex + result.bytesProduced());
+                return result;
+            }
+        },
+        JDK(false, MERGE_CUMULATOR) {
+            @Override
+            SSLEngineResult unwrap(SslHandler handler, ByteBuf in,
+                int readerIndex, int len,
```
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/SslHandler.java]: nit: can you align the parameters ... was a bit hard to look at on first glance

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java [hunk 2]
```diff
@@ -179,8 +179,89 @@
...
+                     * If {@link OpenSslEngine} is in use,
+                     * we can use a special {@link OpenSslEngine#unwrap(ByteBuffer[], ByteBuffer[])} method
+                     * that accepts multiple {@link ByteBuffer}s without additional memory copies.
+                     */
+                    OpenSslEngine opensslEngine = (OpenSslEngine) handler.engine;
+                    try {
+                        handler.singleBuffer[0] = toByteBuffer(out, writerIndex,
+                            out.writableBytes());
+                        result = opensslEngine
+                            .unwrap(in.nioBuffers(readerIndex, len), handler.singleBuffer);
```
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/SslHandler.java]: nit: put this on the previous line?

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java [hunk 3]
```diff
@@ -179,8 +179,89 @@
...
+            return JDK;
+        }
+
+        SslEngineType(boolean wantsDirectBuffer, Cumulator cumulator) {
+            this.wantsDirectBuffer = wantsDirectBuffer;
+            this.cumulator = cumulator;
+        }
+
+        abstract SSLEngineResult unwrap(SslHandler handler,
+            ByteBuf in, int readerIndex, int len, ByteBuf out) throws SSLException;
```
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/SslHandler.java]: nit: fix formatting here too

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java (hunk 1 above)
Comment (nmittler) [inline handler/src/main/java/io/netty/handler/ssl/SslHandler.java]: done

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java (hunk 3 above)
Comment (nmittler) [inline handler/src/main/java/io/netty/handler/ssl/SslHandler.java]: done.

**Comment by nmittler** [comment]
@Scottmitch PTAL

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java [hunk 4]
```diff
@@ -179,8 +179,87 @@
...
+            }
+            return JDK;
+        }
+
+        SslEngineType(boolean wantsDirectBuffer, Cumulator cumulator) {
+            this.wantsDirectBuffer = wantsDirectBuffer;
+            this.cumulator = cumulator;
+        }
+
+        abstract SSLEngineResult unwrap(SslHandler handler, ByteBuf in, int readerIndex, int len, ByteBuf out)
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/SslHandler.java]: @nmittler just one more comment... what you think about also moving the wrap(...) method in there just to make it consistent ?

Comment (nmittler) [inline handler/src/main/java/io/netty/handler/ssl/SslHandler.java]: @normanmaurer I thought about it, but moving wrap into the enum isn't really as helpful. The only parameter to the process is `wantsDirectBuffer` and that's already in the enum. Let's leave it as-is for now. If we find a need to move wrap later, we can. Sound good?

Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/SslHandler.java]: ok... please merge :)

**Comment by normanmaurer** [comment]
@nmittler I squashed and cherry-picked for you into 4.1 (9725a4d004f10dc60d0406c197e1583f91461341) and 4.0 (5a8a365d1740234d6942a218edf5d79f1d8da800)

**Comment by nmittler** [comment]
@normanmaurer great! Thanks, buddy!

**Pull request closed without merging**